pkg/backup/backed_up_items_map.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+// backedUpItemsMap keeps track of the items already backed up for the current Velero Backup
+type backedUpItemsMap struct {
+	*sync.RWMutex
+	BackedUpItems map[itemKey]struct{}
+}
+
+func NewBackedUpItemsMap() *backedUpItemsMap {
+	return &backedUpItemsMap{
+		RWMutex:       &sync.RWMutex{},
+		BackedUpItems: make(map[itemKey]struct{}),
+	}
+}
+
+func (m *backedUpItemsMap) CopyItemList() map[itemKey]struct{} {
+	m.RLock()
+	defer m.RUnlock()
+	returnMap := map[itemKey]struct{}{}
+	for key, val := range m.BackedUpItems {
+		returnMap[key] = val
+	}
+	return returnMap
+}
+
+func (m *backedUpItemsMap) ResourceList() map[string][]string {
+	m.RLock()
+	defer m.RUnlock()
+
+	resources := map[string][]string{}
+	for i := range m.BackedUpItems {
+		entry := i.name
+		if i.namespace != "" {
+			entry = fmt.Sprintf("%s/%s", i.namespace, i.name)
+		}
+		resources[i.resource] = append(resources[i.resource], entry)
+	}
+
+	// sort namespace/name entries for each GVK
+	for _, v := range resources {
+		sort.Strings(v)
+	}
+
+	return resources
+}
+
+func (m *backedUpItemsMap) Size() int {
+	m.RLock()
+	defer m.RUnlock()
+	return len(m.BackedUpItems)
+}
+
+func (m *backedUpItemsMap) ItemInBackup(key itemKey) bool {
```

Comment:
I am fine with the current names. But just thought of `ItemInBackup` <-> `IsItemBackedup` and `AddItem` <-> `MarkItemBackedup`